File: Leetcode/1026_Maximum_Difference_Between_Node_and_Ancestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def dfs(self, node, minval, maxval):
        if node == None: return

        self.ans = max(self.ans, abs(maxval - node.val))
        self.ans = max(self.ans, abs(minval - node.val))

        maxval = max(maxval, node.val)
        minval = min(minval, node.val)
        self.dfs(node.left, minval, maxval)
        self.dfs(node.right, minval, maxval)

    def maxAncestorDiff(self, root) -> int:
        self.ans = 0
        self.dfs(root, root.val, root.val)
        
        return self.ans

File: Leetcode/1026_Maximum_Difference_Between_Node_and_Ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans = 0;
    void dfs(TreeNode* node, int minv, int maxv) {
        minv = min(minv, node->val);
        maxv = max(maxv, node->val);
        ans = max(ans, maxv - minv);
        if (node->left != NULL) dfs(node->left, minv, maxv);
        if (node->right != NULL) dfs(node->right, minv, maxv);
    }
    int maxAncestorDiff(TreeNode* root) {
        dfs(root, root->val, root->val);
        return ans;
    }
};